### PR TITLE
[RHPAM-2813] Allow to use 'user' query parameter with /contents/output REST API.

### PR DESCRIPTION
**Pull request by elguardian**
added user parameter to user save content operation

### CODE REVIEW

**Comment by elguardian** [comment]
depends on https://github.com/kiegroup/jbpm/pull/1628

**Comment by sutaakar** [comment]
@elguardian Please update the revapi-config.json in kie-server-client according to the output of jenkins job.

**Comment by elguardian** [comment]
@afalhambra should work now plus revisited your comments.

File: kie-server-parent/kie-server-services/kie-server-services-jbpm/src/main/java/org/kie/server/services/jbpm/UserTaskServiceBase.java [hunk 1]
```diff
@@ -258,12 +256,17 @@ public void setDescription(String containerId, Number taskId, String description
...
+
+    public String saveContent(String containerId, String userId, Number taskId, String payload, String marshallingType) {
+        userId = getUser(userId);
         containerId = context.getContainerId(containerId, new ByTaskIdContainerLocator(taskId.longValue()));
         logger.debug("About to unmarshal task content parameters from payload: '{}'", payload);
         Map<String, Object> parameters = marshallerHelper.unmarshal(containerId, payload, marshallingType, Map.class);
 
         logger.debug("About to set content of a task with id '{}' with data {}", taskId, parameters);
-        Long contentId = userTaskService.saveContent(containerId, taskId.longValue(), parameters);
+        Long contentId = userTaskService.saveContentFromUser(taskId.longValue(), userId, parameters);
```
Comment (afalhambra) [inline kie-server-parent/kie-server-services/kie-server-services-jbpm/src/main/java/org/kie/server/services/jbpm/UserTaskServiceBase.java:269]: Not really sure whether we need to pass "containerId" as it was done before? Why it is not required now from a functional point of view?

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/ByPassUserTaskServiceIntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    protected void addExtraCustomClasses(Map<String, Class<?>> extraClasses) throws Exception {
+        extraClasses.put(PERSON_CLASS_NAME, Class.forName(PERSON_CLASS_NAME, true, kieContainer.getClassLoader()));
+    }
+
+    @Test
+    @Category(Smoke.class)
+    public void testProcessWithUserTasks() throws Exception {
+        Long processInstanceId = processClient.startProcess(CONTAINER_ID, PROCESS_ID_USERTASK);
+        assertNotNull(processInstanceId);
+        assertTrue(processInstanceId.longValue() > 0);
```
Comment (afalhambra) [inline kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/ByPassUserTaskServiceIntegrationTest.java:66]: unboxing not needed

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/ByPassUserTaskServiceIntegrationTest.java [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+        createContainer(CONTAINER_ID, releaseId);
+    }
+
+    @Override
+    protected void addExtraCustomClasses(Map<String, Class<?>> extraClasses) throws Exception {
+        extraClasses.put(PERSON_CLASS_NAME, Class.forName(PERSON_CLASS_NAME, true, kieContainer.getClassLoader()));
+    }
+
+    @Test
+    @Category(Smoke.class)
```
Comment (afalhambra) [inline kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/ByPassUserTaskServiceIntegrationTest.java:62]: Why is it marked as Smoke category? couldn't find any test suite to execute this category

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/ByPassUserTaskServiceIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+            Map<String, Object> taskOutcome = new HashMap<String, Object>();
+            taskOutcome.put("string_", "my custom data");
+            taskOutcome.put("person_", createPersonInstance(USER_MARY));
+
+            Map<String, Object> content = new HashMap<>();
+            content.put("name", "joda");
+            taskClient.saveTaskContent(CONTAINER_ID, taskSummary.getId(), "admin", content);
+            List<TaskEventInstance> events = taskClient.findTaskEvents(taskSummary.getId(), 0, 10);
+            assertEquals("admin", events.get(events.size() - 1).getUserId());
```
Comment (afalhambra) [inline kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/ByPassUserTaskServiceIntegrationTest.java:86]: not sure why "events.size() - 1"? shouldn't we get always events.size = 1? if not, why events.size - 1 would return "admin" as userId?

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/ByPassUserTaskServiceIntegrationTest.java [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+            taskClient.saveTaskContent(CONTAINER_ID, taskSummary.getId(), "other", content);
+            events = taskClient.findTaskEvents(taskSummary.getId(), 0, 10);
+            assertEquals("other", events.get(events.size() - 1).getUserId());
```
Comment (afalhambra) [inline kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/ByPassUserTaskServiceIntegrationTest.java:90]: same as previous comment

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/ByPassUserTaskServiceIntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+    }
+
+    @Override
+    protected void addExtraCustomClasses(Map<String, Class<?>> extraClasses) throws Exception {
+        extraClasses.put(PERSON_CLASS_NAME, Class.forName(PERSON_CLASS_NAME, true, kieContainer.getClassLoader()));
+    }
+
+    @Test
+    @Category(Smoke.class)
+    public void testProcessWithUserTasks() throws Exception {
```
Comment (afalhambra) [inline kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/ByPassUserTaskServiceIntegrationTest.java:63]: this method doesn't throw any exception

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/ByPassUserTaskServiceIntegrationTest.java [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+        assertEquals(1, task.getPotentialOwners().size());
+        assertEquals(potentialOwner, task.getPotentialOwners().get(0));
+    }
+
+    private void checkTaskStatusAndActualOwner(String containerId, Long taskId, Status status, String actualOwner) {
+        TaskInstance task = taskClient.getTaskInstance(containerId, taskId);
+        assertEquals(taskId, task.getId());
+        assertEquals(status.toString(), task.getStatus());
+        assertEquals(actualOwner, task.getActualOwner());
+    }
```
Comment (afalhambra) [inline kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/ByPassUserTaskServiceIntegrationTest.java:146]: These 4 methods are never used

File: kie-server-parent/kie-server-remote/kie-server-client/src/main/java/org/kie/server/client/UserTaskServicesClient.java [hunk 8]
```diff
@@ -73,6 +73,8 @@
 
     Long saveTaskContent(String containerId, Long taskId, Map<String, Object> values);
 
+    Long saveTaskContent(String containerId, Long id, String userId, Map<String, Object> content);
```
[the remark this answers is not in the document]
Comment (elguardian) [reply kie-server-parent/kie-server-remote/kie-server-client/src/main/java/org/kie/server/client/UserTaskServicesClient.java:76]: done

File: kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-jbpm/src/main/java/org/kie/server/remote/rest/jbpm/UserTaskResource.java [hunk 9]
```diff
@@ -707,6 +707,7 @@ public Response setDescription(@Context HttpHeaders headers,
     public Response saveContent(@Context HttpHeaders headers, 
             @ApiParam(value = "container id that task instance belongs to", required = true, example = "evaluation_1.0.0-SNAPSHOT") @PathParam(CONTAINER_ID) String containerId,
             @ApiParam(value = "identifier of the task instance that data should be saved into", required = true, example = "123") @PathParam(TASK_INSTANCE_ID) Long taskId, 
+                                @ApiParam(value = "optional user id to be used instead of authenticated user - only when bypass authenticated user is enabled", required = false) @QueryParam("user") String userId,
```
[the remark this answers is not in the document]
Comment (elguardian) [reply kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-jbpm/src/main/java/org/kie/server/remote/rest/jbpm/UserTaskResource.java:710]: just left it

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/ByPassUserTaskServiceIntegrationTest.java (hunk 2 above)
Comment (elguardian) [reply kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/ByPassUserTaskServiceIntegrationTest.java:66]: ok

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/ByPassUserTaskServiceIntegrationTest.java (hunk 3 above)
Comment (elguardian) [reply kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/ByPassUserTaskServiceIntegrationTest.java:62]: removed

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/ByPassUserTaskServiceIntegrationTest.java (hunk 4 above)
Comment (elguardian) [reply kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/ByPassUserTaskServiceIntegrationTest.java:86]: it is the lasts event. We are checking that the last event modification is by the user being impersonated.

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/ByPassUserTaskServiceIntegrationTest.java (hunk 5 above)
Comment (elguardian) [reply kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/ByPassUserTaskServiceIntegrationTest.java:90]: checking the last event is by the user being impersonated.

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/ByPassUserTaskServiceIntegrationTest.java (hunk 6 above)
Comment (elguardian) [reply kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/ByPassUserTaskServiceIntegrationTest.java:63]: removed

File: kie-server-parent/kie-server-services/kie-server-services-jbpm/src/main/java/org/kie/server/services/jbpm/UserTaskServiceBase.java (hunk 1 above)
Comment (elguardian) [reply kie-server-parent/kie-server-services/kie-server-services-jbpm/src/main/java/org/kie/server/services/jbpm/UserTaskServiceBase.java:269]: not really needed when you have the task id

**Comment by elguardian** [comment]
Jenkins retest this please.

**Comment by elguardian** [comment]
this is ok to test.

**Comment by elguardian** [comment]
this is ok to test.

**Comment by elguardian** [comment]
this is ok to test.

**Comment by elguardian** [comment]
Jenkins retest this please.

**Comment by elguardian** [comment]
@afalhambra can you review ?

**Comment by elguardian** [comment]
@afalhambra @gmunozfe can you review this please.

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/pom.xml [hunk 10]
```diff
@@ -268,6 +268,7 @@
           <configuration>
             <systemPropertyVariables>
               <!-- Values are supplied by specific profiles -->
+              <org.kie.server.bypass.auth.user>true</org.kie.server.bypass.auth.user>
```
Comment (afalhambra) [inline kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/pom.xml:271]: Not sure why the property is specified in the cargo plugin?

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/pom.xml [hunk 11]
```diff
@@ -326,6 +327,7 @@
         <configuration>
           <container>
             <systemProperties>
+              <org.kie.server.bypass.auth.user>true</org.kie.server.bypass.auth.user>
```
Comment (afalhambra) [inline kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/pom.xml:330]: Same here. Why is the property specified in the integration test plugin?

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/RollingUpdateUserTaskServiceIntegrationTest.java [hunk 12]
```diff
@@ -64,7 +64,7 @@ public void testGetTaskInputAndOutputWithAlias() throws Exception {
             ProcessInstance processInstance = processClient.getProcessInstance(CONTAINER_ALIAS, pid);
             assertThat(processInstance).isNotNull();
 
-            List<TaskSummary> taskList = taskClient.findTasksAssignedAsBusinessAdministrator(USER_YODA, 0, 10);
+            List<TaskSummary> taskList = taskClient.findTasksAssignedAsBusinessAdministrator(USER_ADMINISTRATOR, 0, 10);
```
Comment (afalhambra) [inline kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/RollingUpdateUserTaskServiceIntegrationTest.java:67]: Why this change?

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/pom.xml (hunk 10 above)
Comment (elguardian) [reply kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/pom.xml:271]: it is requires for the server to enable by pass user. Otherwise does not work.

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/pom.xml (hunk 11 above)
Comment (elguardian) [reply kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/pom.xml:330]: same as with cargo. this is for enabling client bypass client

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/RollingUpdateUserTaskServiceIntegrationTest.java (hunk 12 above)
Comment (elguardian) [reply kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/RollingUpdateUserTaskServiceIntegrationTest.java:67]: this is a faulty test.  before it was working because because the bypass was on false. But when I changed it the the administrator user was the right one to set.
